stonesoup/sensormanager/reward.py

```diff
@@ -233,44 +244,62 @@ def __call__(self, config: Mapping[Sensor, Sequence[Action]], tracks: Set[Track]
+        det_count = 0
+        for sensor, detections in sensor_detections.items():
 
             for predicted_track, detection_set in detections.items():
-
+                det_count += len(detection_set)
                 for n, detection in enumerate(detection_set):
 
-                    # if detection:
                     # Generate hypothesis based on prediction/previous update and detection
                     hypothesis = SingleHypothesis(predicted_track, detection)
 
                     # Do the update based on this hypothesis and store covariance matrix
                     update = self.updater.update(hypothesis)
 
-                    # else:
-                    #     update = copy.copy(predicted_track[-1])
-
                     kld += self.KLD(predicted_track[-1], update)
 
-            if self.method_sum is False and len(detections) != 0:
-
-                kld /= len(detections)
+                    if not isinstance(self, MultiUpdateExpectedKLDivergence):
+                        predicted_track.append(update)
 
-        # Return value of configuration metric
-        return kld
+        if self.method_sum is False and det_count != 0:
 
-    def _generate_detections(self, predicted_tracks, sensor):
+            kld /= det_count
 
-        detections = {}
-        for predicted_track in predicted_tracks:
-            track_detections = set()
-            track_detections.update(sensor.measure({State(predicted_track.mean)}, noise=True))
-
-            detections.update({predicted_track: track_detections})
-
-        return detections
+        # Return value of configuration metric
+        if self.return_tracks:
+            return kld, predicted_tracks
+        else:
+            return kld
+
+    def _generate_detections(self, predicted_tracks, sensors, timestamp=None):
+
+        all_detections = {}
+        detection_set = set()
+
+        for sensor in sensors:
+            detections = {}
+            for predicted_track in predicted_tracks:
+                tmp_detection = sensor.measure({State(predicted_track.mean,
+                                                      timestamp=predicted_track.timestamp)},
+                                               noise=True)
+                detection_set |= tmp_detection
+                if self.data_associator:
+                    tmp_hypotheses = self.data_associator.associate(predicted_tracks,
+                                                                    tmp_detection,
+                                                                    timestamp)
+                    detections.update({predicted_track: {hypothesis.measurement}
+                                      for predicted_track, hypothesis in tmp_hypotheses.items() if
+                                      hypothesis})
+                else:
+                    detections.update({predicted_track: tmp_detection})
+
+            all_detections.update({sensor: detections})
+
+        return all_detections
```

Comment:
I think this may be better if you considered all the tracks and all the detections from a single sensor in one data association step.

```diff
diff --git a/stonesoup/sensormanager/reward.py b/stonesoup/sensormanager/reward.py
index af40d935..a6566920 100644
--- a/stonesoup/sensormanager/reward.py
+++ b/stonesoup/sensormanager/reward.py
@@ -278,7 +278,6 @@ class ExpectedKLDivergence(RewardFunction):
     def _generate_detections(self, predicted_tracks, sensors, timestamp=None):
 
         all_detections = {}
-        detection_set = set()
 
         for sensor in sensors:
             detections = {}
@@ -286,16 +285,16 @@ class ExpectedKLDivergence(RewardFunction):
                 tmp_detection = sensor.measure({State(predicted_track.mean,
                                                       timestamp=predicted_track.timestamp)},
                                                noise=True)
-                detection_set |= tmp_detection
-                if self.data_associator:
-                    tmp_hypotheses = self.data_associator.associate(predicted_tracks,
-                                                                    tmp_detection,
-                                                                    timestamp)
-                    detections.update({predicted_track: {hypothesis.measurement}
-                                      for predicted_track, hypothesis in tmp_hypotheses.items() if
-                                      hypothesis})
-                else:
-                    detections.update({predicted_track: tmp_detection})
+                detections.update({predicted_track: tmp_detection})
+
+            if self.data_associator:
+                tmp_hypotheses = self.data_associator.associate(
+                    predicted_tracks,
+                    {det for dets in detections.values() for det in dets},
+                    timestamp)
+                detections = {predicted_track: {hypothesis.measurement}
+                              for predicted_track, hypothesis in tmp_hypotheses.items()
+                              if hypothesis}
 
             all_detections.update({sensor: detections})
 

 ```

Reply:
That is probably a better way to do it. I was thinking of not overwriting if we could get away with it but that will mean we are only providing a single track object